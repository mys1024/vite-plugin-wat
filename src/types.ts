export interface VitePluginWatOptions {
  /**
   * Output directory of the generated WASM files. (default: 'node_modules/.vite-plugin-wat/wasm')
   */
  wasmDir?: string
}
