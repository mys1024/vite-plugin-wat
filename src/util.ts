export class TimingMap<K, V> {
  #map = new Map<K, V>()
  #timeout = new Map<K, number | NodeJS.Timeout>()

  size(): number {
    return this.#map.size
  }

  has(key: K): boolean {
    return this.#map.has(key)
  }

  get(key: K): V | undefined {
    return this.#map.get(key)
  }

  set(key: K, value: V, survivalTime: number) {
    const oldTimeout = this.#timeout.get(key)
    if (oldTimeout !== undefined)
      clearTimeout(oldTimeout)
    const timeout = setTimeout(() => {
      this.#map.delete(key)
      this.#timeout.delete(key)
    }, survivalTime)
    this.#map.set(key, value)
    this.#timeout.set(key, timeout)
  }
}
