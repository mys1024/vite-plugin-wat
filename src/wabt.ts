import wabt from 'wabt'

export async function wat2wasm(wat: string) {
  const wabtModule = await wabt()
  const wasmModule = wabtModule.parseWat('', wat)
  const wasm = wasmModule.toBinary({}).buffer
  wasmModule.destroy()
  return wasm
}
