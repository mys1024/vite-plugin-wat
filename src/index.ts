import type { Plugin } from 'vite'
import { relative, resolve } from 'pathe'
import fse from 'fs-extra'

import { wat2wasm } from './wabt'
import { TimingMap } from './util'

const PLUGIN_NAME = 'vite-plugin-wat'
const WAT_ID_REG = /\.wat\?init$/

function isWatId(id: string): boolean {
  return WAT_ID_REG.test(id)
}

function resolveWasmId(watId: string, cwd: string, wasmDir: string) {
  return resolve(wasmDir, relative(cwd, watId).replace(WAT_ID_REG, '.wasm'))
}

export default (): Plugin => {
  const cwd = resolve('./')
  const storeDir = resolve(cwd, 'node_modules', `.${PLUGIN_NAME}`)
  const wasmDir = resolve(storeDir, 'wasm')
  const transformCache = new TimingMap<
    string,
    { code: string; output: string }
  >()

  return {
    name: PLUGIN_NAME,
    config: () => ({
      server: { watch: { ignored: [wasmDir] } },
    }),
    buildStart: () => {
      fse.removeSync(wasmDir)
    },
    resolveId: (id) => {
      return isWatId(id) ? id : undefined
    },
    transform: async (code, id) => {
      if (!isWatId(id))
        return
      // check cache
      const cache = transformCache.get(id)
      if (cache?.code === code)
        return cache.output
      // wasmId will be a string like '/path/to/store/wasm/generated.wasm'
      const wasmId = resolveWasmId(id, cwd, wasmDir)
      // compile `.wat` to `.wasm`, save `.wasm` in `wasmDir`
      await fse.ensureFile(wasmId)
      await fse.writeFile(wasmId, await wat2wasm(code))
      // transform to export from '*.wasm?init'
      const output = `export { default } from '${wasmId}?init';`
      transformCache.set(id, { code, output }, 1000 * 60 * 10) // cache will expire after 10 min
      return output
    },
  }
}
