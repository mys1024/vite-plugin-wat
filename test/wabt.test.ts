import { describe, expect, it } from 'vitest'
import { wat2wasm } from '../src/wabt'

describe('wat2wasm', () => {
  const wat = `
    (module
      (func $add (param i32) (param i32) (result i32)
        local.get 0
        local.get 1
        i32.add
      )
      (export "add" (func $add))
    )
  `

  it('compiling', async () => {
    const wasm = await wat2wasm(wat)
    expect(wasm).toMatchSnapshot()
  })

  it('correctness', async () => {
    const wasm = await wat2wasm(wat)
    const { add } = (
      await (WebAssembly.instantiate(wasm, {}))
    ).instance.exports
    if (add instanceof Function)
      expect(add(1, 2)).toBe(3)
    else
      throw new Error('\'add\' is not a function')
  })
})
