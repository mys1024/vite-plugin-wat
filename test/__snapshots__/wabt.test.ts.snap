// Vitest Snapshot v1

exports[`wat2wasm > compiling 1`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  1,
  96,
  2,
  127,
  127,
  1,
  127,
  3,
  2,
  1,
  0,
  7,
  7,
  1,
  3,
  97,
  100,
  100,
  0,
  0,
  10,
  9,
  1,
  7,
  0,
  32,
  0,
  32,
  1,
  106,
  11,
]
`;
